### PR TITLE
NFTGallery and NFT

### PR DESCRIPTION
**Pull request by hariria**
Updated NFTGallery and NFT from the `trove-hackathon` branch, changed `assert` to `assert!` in NFT.move and NFTGallery.move

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Ran:
```bash
cargo x clippy && cargo x fmt
```
```bash
cargo x build && cargo run --release -- --package experimental
```
```bash
cargo x test
```

### CODE REVIEW

**Comment by gregnazario** [comment]
There are some whitespace issues stopping CI/CD btw.
```
Found whitespace errors in commit 5518c9fed8320e880735d60b9a8216199e2493bf:

diem-move/diem-framework/experimental/releases/artifacts/current/build/DiemExperimental/sources/NFTGallery.move:122: new blank line at EOF.
diem-move/diem-framework/experimental/sources/NFTGallery.move:122: new blank line at EOF.
```

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (gregnazario) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:14]: I think I commented this on someone else's PR (not necessarily yours to fix).  We should probably just have this built into the language.

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (gregnazario) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:27]: This seems weird, one line? (or at like the last function in this file).  Not sure if we have a formatter for move.

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (gregnazario) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:42]: Similar

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (gregnazario) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:69]: Similar

File: diem-move/diem-framework/experimental/sources/NFT.move
Comment (gregnazario) [inline diem-move/diem-framework/experimental/sources/NFT.move:144]: There's a lot of duplication in here.  We should later (not in this PR), consider either removing the deduplication, or asking the move team to provide these kinds of functionalities (e.g. iterate to find token or even just for loops)

**Review by gregnazario — COMMENTED** [review]
Looks like you've definitely accomplished what we want.  I just want to make sure we have someone who's a little more up to date on their move double check the code how it is @emmazzz could you take a look and ensure it's all good (since we didn't have anyone really review it earlier)

**Comment by hariria** [comment]
Ran
```bash
cargo x fmt && cargo x lint && cargo x clippy -p diem-framework
```
also re-ran
```bash
cargo x build && cargo run --release -- --package experimental
```

**Comment by dmitri-perelman** [comment]
Thanks @hariria !
Could you also include the tests from [here](https://github.com/diem/diem/blob/trove-hackathon/language/diem-framework/tests/experimental/NFTTests.move)?

The right place for the tests would be under `diem-move/diem-framework/experimental/tests`.
Just to verify (since it's not obvious):
* In order to build the package you want to run `cargo run --package experimental` from `diem-move/diem-framework`
* In order to run the unit tests you want to run `df-cli package test --instructions=50000` from `diem-move/diem-framework/experimental`
* You're not going to update the demo-cli in this PR because you're not adding any script functions here yet.

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (dmitri-perelman) [reply diem-move/diem-framework/experimental/sources/NFTGallery.move:14]: I think it's actually not used here at all, remove?

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (dmitri-perelman) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:33]: Is there a reason we can't merge the new token with the original token in place? I thought we could `borrow_mut` and `NFT::join`. @emmazzz is it something I'm missing?

File: diem-move/diem-framework/experimental/sources/NFTGallery.move
Comment (dmitri-perelman) [inline diem-move/diem-framework/experimental/sources/NFTGallery.move:97]: Same here: I think it's going to be pretty inefficient since we're doing unnecessary `Vector::remove()` and `Vector::push_back()` operations.

**Review by dmitri-perelman — COMMENTED** [review]
Some questions inlined.
I think we should not stop this PR from landing and iterate on the implementation once it's merged.

**Comment by hariria** [comment]
**Ran from `diem-move > diem-framework > experimental`**
```
cargo run -p df-cli -- package test --instructions=50000  
```

**also re-ran**
```bash
cargo run --release -- --package experimental
```
```bash
 cargo run -p diem-framework --release
```
```bash
cargo x fmt && cargo x lint && cargo x clippy -p diem-framework
```

**Review by gregnazario — APPROVED** [review]
Let's move fast and circle back on the comments afterwards.  Looks like the test failures earlier are unrelated.

/land

**Pull request closed without merging**
